File: app/src/main/java/com/util/FinalArray.java
package com.util;

import android.support.annotation.NonNull;

import java.util.Arrays;
import java.util.Iterator;
import java.util.NoSuchElementException;

/**
 * A FinalArray is an array that is immutable.
 */
public class FinalArray<T> implements Iterable<T> {
    private T[] arr;

    private FinalArray(T[] arr) {
        this.arr = Arrays.copyOf(arr, arr.length);
    }

    public static <T> FinalArray<T> from(T[] arr) {
        return new FinalArray<>(arr);
    }

    @SuppressWarnings("unused")
    public T get(int index) {
        return arr[index];
    }

    @Override
    public String toString() {
        return Arrays.toString(arr);
    }

    @Override @NonNull
    public Iterator<T> iterator() {
        return new Iterator<T>() {
            private int current = 0;

            @Override
            public boolean hasNext() {
                return current < FinalArray.this.arr.length;
            }

            @Override
            public T next() {
                if (!hasNext()) {
                    throw new NoSuchElementException();
                }
                return arr[current++];
            }
        };
    }
}

File: app/src/main/java/com/util/FinalArray.kt
package com.util

import java.util.Arrays
import java.util.NoSuchElementException

@Suppress("unused")
/**
 * A FinalArray is an array that is immutable.
 *
 * It only supports arrays of objects though, so convert arrays of primitive types
 * into their respective wrapper classes first.
 * For example, passing an int[] would give a compiler error,
 * but passing an Integer[] wouldn't.
 */
class FinalArray<T> private constructor(arr: Array<T>) : Iterable<T> {
    private val arr: Array<T> = Arrays.copyOf(arr, arr.size)

    operator fun get(index: Int): T {
        return arr[index]
    }

    fun length(): Int {
        return arr.size
    }

    override fun toString(): String {
        return Arrays.toString(arr)
    }

    override fun iterator(): Iterator<T> {
        return object : Iterator<T> {
            private var current = 0

            override fun hasNext(): Boolean {
                return current < this@FinalArray.arr.size
            }

            override fun next(): T {
                if (!hasNext()) {
                    throw NoSuchElementException()
                }
                return arr[current++]
            }
        }
    }

    companion object {
        fun <T> from(arr: Array<T>): FinalArray<T> {
            return FinalArray(arr)
        }
    }
}